AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/Session/ActivityTracking/ActivityTracker.swift

```diff
@@ -50,16 +55,123 @@ protocol ActivityTrackerBehaviour {
-#if !os(OSX)
-        return UIActivityTracker(backgroundTrackingTimeout: timeout)
-#else
-        return EmptyTracker()
+class ActivityTracker: ActivityTrackerBehaviour {
+
+#if canImport(UIKit)
+    private var backgroundTask: UIBackgroundTaskIdentifier = .invalid
 #endif
+
+    private var backgroundTimer: Timer? {
+        willSet {
+            backgroundTimer?.invalidate()
+        }
+    }
+    private let backgroundTrackingTimeout: TimeInterval
+    private let stateMachine: StateMachine<ApplicationState, ActivityEvent>
+    private var stateMachineSubscriberToken: StateMachineSubscriberToken?
+
+    private static let applicationDidMoveToBackgroundNotification: Notification.Name = {
+    #if canImport(UIKit)
+        UIApplication.didEnterBackgroundNotification
+    #else
+        NSApplication.didResignActiveNotification
+    #endif
+    }()
+
+    private static  let applicationWillMoveToForegoundNotification: Notification.Name = {
+    #if canImport(UIKit)
+        UIApplication.willEnterForegroundNotification
+    #else
+        NSApplication.willBecomeActiveNotification
+    #endif
+    }()
+
+    private static  var applicationWillTerminateNotification: Notification.Name = {
+    #if canImport(UIKit)
+        UIApplication.willTerminateNotification
+    #else
+        NSApplication.willTerminateNotification
+    #endif
+    }()
+
+    private let notifications = [
+        applicationDidMoveToBackgroundNotification,
+        applicationWillMoveToForegoundNotification,
+        applicationWillTerminateNotification
+    ]
+
+    init(backgroundTrackingTimeout: TimeInterval,
+         stateMachine: StateMachine<ApplicationState, ActivityEvent>? = nil) {
+        self.backgroundTrackingTimeout = backgroundTrackingTimeout
+        self.stateMachine = stateMachine ?? StateMachine(initialState: .initializing,
+                                                         resolver: ApplicationState.Resolver.resolve(currentState:event:))
+        for notification in notifications {
+            NotificationCenter.default.addObserver(self,
+                                                   selector: #selector(handleApplicationStateChange),
+                                                   name: notification,
+                                                   object: nil)
+        }
+    }
+
+    deinit {
+        for notification in notifications {
+            NotificationCenter.default.removeObserver(self,
+                                                      name: notification,
+                                                      object: nil)
+        }
     }
 
-    private class EmptyTracker: ActivityTrackerBehaviour {
-        func beginActivityTracking(_ listener: @escaping (ApplicationState) -> Void) {}
+    func beginActivityTracking(_ listener: @escaping (ApplicationState) -> Void) {
+        stateMachineSubscriberToken = stateMachine.subscribe(listener)
```

Comment:
set `stateMachineSubscriberToken` to nil in deinit?

Reply:
oops, will do